source/tilecameragenerator.ts

```diff
@@ -0,0 +1,491 @@
+import { mat4, vec3, vec4 } from 'gl-matrix';
+import { log, LogLevel } from './auxiliaries';
+import { Camera } from './camera';
+import { m4 } from './gl-matrix-extensions';
+
+/**
+ * Support Class that wraps the calculation of camera tiling and iteration with various algorithms
+ * by giving access to a adjusted camera which NDC-Coordinates match the current tile index.
+ * Iteration can be done manually (variant: 1) or automatically (variant: 2).
+ * It is intended to be used like:
+ *
+ * // setup
+ * tileCameraGenerator = new TileCameraGenerator();
+ * tileCameraGenerator.sourceCamera = camera;
+ * tileCameraGenerator.sourceViewPort = canvSize;
+ * tileCameraGenerator.tileSize = [128, 128];
+ * tileCameraGenerator.padding = new vec4();
+ * tileCamera = tileRender.camera;
+ * tileCameraGenerator.algorithm = TileCameraGenerator.IteratorAlgorithm.ScanLine;
+ * let offset: [number, number];
+ *
+ * iteration variant 1:
+ * for (let i = 0; i < tileCameraGenerator.numberOfTiles(); ++i){
+ *      tileCameraGenerator.tile = i; // property
+ *      tileCameraGenerator.update();
+ *      offset = tileCameraGenerator.offset;
+ *      "render camera"
+ * }
+ *
+ * iteration variant 2:
+ * while(tileCameraGenerator.nextTile()){
+ *      offset = tileCameraGenerator.offset;
+ *      "render"
+ * }
+ * // reset generator
+ * tileCameraGenerator.resetTileRendering();
+ *
+ * NOTE: Use updateCameraProperties if the source camera is altered.
+ */
+
+
+export class TileCameraGenerator {
+
+    /** @see {@link eye} */
+    protected _camera: Camera | undefined;
+
+    /** @see {@link sourceCamera} */
+    protected _sourceCamera: Camera | undefined;
+
+    /** @see {@link tile} */
+    protected _tile = -1;
+
+    /** @see {@link sourceViewPort} */
+    protected _sourceViewPort: [number, number] | undefined;
+
+    /** @see {@link tileSize} */
+    protected _tileSize: [number, number] | undefined;
+
+    /** @see {@link padding} */
+    protected _padding: vec4 = vec4.fromValues(0, 0, 0, 0);
+
+    /** @see {@link algorithm} */
+    protected _algorithm: TileCameraGenerator.IterationAlgorithm = TileCameraGenerator.IterationAlgorithm.ScanLine;
+
+    protected _valid: boolean;
+    protected _currentOffset: [number, number] = [0, 0];
+    protected _iterationAlgorithmIndices: [number, number][] = [];
+
+    /**
+     * Converts the tile index from the selected Algorithm to table indices.
+     * @returns - The converted tile index.
+     */
+    protected tileIndexToTableIndices(): [number, number] {
+        if (this.algorithm === TileCameraGenerator.IterationAlgorithm.ScanLine) {
```

Comment:
Depending on the iteration algorithm, you chose different approaches:
1. In-line computation (scan line)
2. Precomputation and lookup (hilbert curve, z curve)
I suggest to convert all variations to precomputed `_iterationAlgorithmIndices` as it decouples generation logic and access logic. This results in significantly reduced code for access and a clear structure for generation.
One down side: the memory footprint of scan line algorithm slightly increases.

What seems to generally hold is that the contents on `_iterationAlgorithmIndices` solely rely on the algorithm and the width and height of the raster. As they usually don't match with the number of fragments of the canvas, we may even pre-compute the raster for several sizes.